### DISCUSSION TITLE
Old Version Works, latest crashes

### DISCUSSION THREAD

**Discussion by Jay-Ti** [visualapproach/WiFi-remote-for-Bestway-Lay-Z-SPA#398] [Q&A]
Well I'm at a loss here,

I have a 4 wire Miami 54123 with a V1 Board.
Having installed v3.3.1 it ran reasonably fine apart from a few crashes every few days or so.
But installing the Latest version or the dev version (or even 3.4.0) it just keeps on crashing on the Hardware Watchdog.

I've narrowed it down to the CIO. Having it disconnected, The ESP boots and stays online for a long time. 
But connecting the CIO makes it crash Instantly after booting. And does nothing when connecting it after it booted.
As far as I've seen the pinouts haven't changed, so since it worked on 3.3.1 i assume my pinout is correct for the latest versions and hasn't changed since it worked on 3.3.1. Especially since the DSP is working.

I have cleared the flash, I've recompiled and re-uploaded everything including a new file system, i got the same settings defined the V1 as the board.

Any ideas or help to get the latest Versions running would be Highly appreciated.

**Comment by visualapproach** [comment]
Turn off mqtt. Close all browsers. Update from usb (occasionally I get weird reset loops when updating via OTA)

**Reply by Jay-Ti** [reply]
I always update over USB.
Although got some instability at some point because of that.
Turning off mqtt would be the death of the project for me.
It doesnt reach anything mqtt related in the setup iirc. But ill look what it does before crashing when i have the time the next few days.

**Reply by visualapproach** [reply]
It was not as the ultimate solution. More to see if it makes a difference. But check where it goes bad and we go from there.

**Reply by Jay-Ti** [reply]
sometimes crashes at bwc.loop
sometimes crahses at server.handle 
sometimes crashes at OTA.handle
sometimes doesnt even wholly write out the Serial println statements i used to check where it crahsed before crashing
all of those functions are in the void loop, so it goes through the stup no problem.
no consistency as far as i can see in when/ where it crashes.
adding wdtFeed doesnt help.
Wifi signal and connection is great.
Setting the CPU frequency to 80MHZ or 160 also doesnt make a difference

**Reply by visualapproach** [reply]
Okay, it may be some interrupt routine that crashes. It may not show in the traces. Maybe you can remove code from cio interrupt and see if it still crashes. It will ofc not function but to pin down the error.

**Reply by Jay-Ti** [reply]
Are there interrupt functions? If so, I can't find them. I only see the Functions in BWC_8266_4w BWC::loop cio::loop etc.

I've narrowed it down to line 209 ESP.wdtEnable(0) in the first bwc::loop after boot.
Got a println statement before and after, and it **almost** consistently stops in the println before the command at the first call of the loop(almost consistently? yep... sometimes crashes at tracking stat changes, the end of the bwc::loop before doing anything else, sometimes at _handleStateChanges() sometimes it even does a handful of loops before crashing at the end of the BWC loop). It doesn't even fully write out the line to the console.

Never disabling the watchdog doesn't work, even when feeding it just before the println,
Re-enabling it with a higher value also doesn't change the behavior.

I find it really odd it doesn't fully write out the line to the serial console. I'll see if the internet has encountered this before.

**Reply by visualapproach** [reply]
Oh yeah, the 4-wire, well there is no ISR in my library there. How long before it crashes after restart?

**Reply by Jay-Ti** [reply]
Most of the time around 1500ms ±100ms according to millis() if the wifi connects instantly.

**Reply by visualapproach** [reply]
Strange. What's the reboot reason according to log? Does it work if you disable mqtt? The major diff from 3.3.1 is HA sensors and maybe that's part of the problem, hence the mqtt question.

**Reply by visualapproach** [reply]
> Never disabling the watchdog doesn't work, even when feeding it just before the println,
Re-enabling it with a higher value also doesn't change the behavior.

what if you feed it just before enabling it?

**Reply by Jay-Ti** [reply]
I mixed up versions.
Current Master works fine, but i installed the dev branch. So the bahaviour is only on the current dev branch. Will try to disable mqtt 

Does feeding before enabling even make sense?
If you disable the sw watchdog it only relies on the hw watchdog which cant be fed?

**Reply by visualapproach** [reply]
> I mixed up versions.
> 
> Current Master works fine, but i installed the dev branch. So the bahaviour is only on the current dev branch. Will try to disable mqtt 
> 
> 
> 
> Does feeding before enabling even make sense?
> 
> If you disable the sw watchdog it only relies on the hw watchdog which cant be fed?

I don't know. You haven't stated the reboot reason yet as far as I can see, but presuming it's a HWWDT? You can compare the two versions (master/dev) on GitHub and see if you can backtrack to a point where it works. The dev version works for me so I can't test.

**Reply by heijjan** [reply]
I just updated my Helsinki to the latest dev version, it works fine for me. The only thing that doesn't work is the ambient temperature update, my sensor gives data in celcius, but for some reason I get a negative temp reading to Home Assistant for the ambient temp value. It seems that something is now thinking that the input-celcius is in Fahrenheit and Home Assistant converts that value back to Celcius. The input value at the moment is 7 celcius and HA reports it as -13 degrees. It worked fine with the couple months old master release.

**Reply by visualapproach** [reply]
I'll reply in the proper thread

**Reply by Jay-Ti** [reply]
The Problem Continues to elude me. 
I got Power without having the USB Connected -> Power Connected ok
I got DSP alive ->  DSP connectors are connected ok
I got 54123 and  PCBV1 defined esp boots fine and stays alive.
Only thing i gotta do is find out is ht RX/TX cable for the CIO.
Put it in one way -> CIO dead, ESP stays alive. 
Put it in the other way -> End of Setup() 8 seconds later boom wdt resets reset cause: 4 all day. 
Mind you everything still works fine if plugged in as intended by the Company, so i didn't fry anything(yet).

been brooding over this on and off for months

**Reply by visualapproach** [reply]
To find out which wire is Rx/Tx from Cio I think you can measure voltage between GND and the pins. No esp or display connected. If you get a reading on one of them, around 4.5V on a DMM, that's the serial out from the CIO's perspective.
My guess is it's the "one way" and not "the other way" that's correct.
You can try defining PCB_V2B and connecting directly to the NodeMCU without the LLC. Connect to the pins according to V2B definitions.
This may solve the potential com problem to the CIO.

**Reply by Jay-Ti** [reply]
Managed to get 3.5.0 working and pretty stable, can control from dsp and Webpage.
Using the same Hardware Setup, But updating To 4.1.1 and uploading the corresponding file system and restarting everything,choosing 54123 in the hardware setup and restarting again,  i get nothing, 04 from the dsp, no control over webpage(cant even toggle god mode).

now, the working setup is cio rx to D3 and tx to D2  , Dsp rx to D6 and tx to D7.
Choosing 54123 for CIO and DSP and V1 board in Hardware config on 4.1.1 flips the DSP pins to rx 7 and tx 6. 
But even using custom pinouts and changing them back with a D, or using the GPIO numbers(and power cycling everything) doesn't lead to anything working.

Am i missing something i haven't tried yet? Or should i just give up and embrace "never touch a running system"?

**Reply by visualapproach** [reply]
The pinout in v3.5 was the same as v4 for PCB v1. I tricked myself too because RX/TX is not in the same order on the CIO and DSP:
const int CIO_RX = D3;
const int CIO_TX = D2;
const int **DSP_TX = D6;**
const int DSP_RX = D7;

The god mode will exit as soon as an error is shown, so if you have E04 you need to reset the pump before god mode can work.
Internet says E04 is overheating error. Could it be overheated or should I see if the ESP reports F instead of C to the display maybe. What if you only use C as a test.

**Reply by Jay-Ti** [reply]
Using the pinout you just described nothing changes. 
04 is not E04, 04 on the dsp Segment display is what you get when there is no communication to the DSP, at least from my two 54123 pumps.
So yeah, no communication between the ESP to the CIO nor the DSP.
Although, is the "Temperature Actual" Value always derived from communication with  the cio? Because i got a value there, i just cant do anything else since God mode doesnt activate.  Also cant change the Temperature Unit if theres no communication :P

**Reply by visualapproach** [reply]
I'm looking at the god mode issue. Temp is initialized to 25C but then derived from CIO messages

**Reply by visualapproach** [reply]
I have fixed a bug in previous beta version. It didn't update messages to the display if not in god mode. I also made it possible to enter god mode without good communication. Plz try update to beta version. (You can do it from the web ui)
